pkg/ugrpc/c1_credential_provider.go

```diff
@@ -0,0 +1,245 @@
+package ugrpc
+
+import (
+	"bytes"
+	"context"
+	"crypto/ed25519"
+	"encoding/base64"
+	"errors"
+	"net"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/conductorone/baton-sdk/pkg/uhttp"
+	"github.com/grpc-ecosystem/go-grpc-middleware/logging/zap/ctxzap"
+	"github.com/pquerna/xjwt"
+	"golang.org/x/oauth2"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/status"
+	"gopkg.in/square/go-jose.v2"
+	"gopkg.in/square/go-jose.v2/json"
+	"gopkg.in/square/go-jose.v2/jwt"
+)
+
+const (
+	assertionType = "urn:ietf:params:oauth:client-assertion-type:jwt-bearer"
+)
+
+var (
+	ErrInvalidClientSecret = errors.New("invalid client secret")
+	ErrInvalidClientID     = errors.New("invalid client id")
+)
+
+type c1Token struct {
+	AccessToken string `json:"access_token"`
+	TokenType   string `json:"token_type"`
+	Expiry      int    `json:"expires_in"`
+}
+
+type c1TokenSource struct {
+	clientID     string
+	clientSecret *jose.JSONWebKey
+	tokenHost    string
+	httpClient   *http.Client
+}
+
+func parseClientID(input string) (string, string, error) {
+	// split the input into 2 parts by @
+	items := strings.SplitN(input, "@", 2)
+	if len(items) != 2 {
+		return "", "", ErrInvalidClientID
+	}
+	clientName := items[0]
+
+	// split the right part into 2 parts by /
+	items = strings.SplitN(items[1], "/", 2)
+	if len(items) != 2 {
+		return "", "", ErrInvalidClientID
+	}
+
+	return clientName, items[0], nil
+}
+
+func parseSecret(input []byte) (*jose.JSONWebKey, error) {
+	items := bytes.SplitN(input, []byte(":"), 4)
+	if len(items) != 4 {
+		return nil, ErrInvalidClientSecret
+	}
+
```

Comment:
missing check if items[2] == "v1"?

Reply:
This is fixed in 8700b1a260f38ac4fd892f0e608c3c4289919ecc